helix-core/src/main/java/org/apache/helix/task/WorkflowContext.java

```diff
@@ -200,23 +222,37 @@ public long getFinishTime() {
   }
 
   public void setLastScheduledSingleWorkflow(String workflow) {
-    _record.setSimpleField(WorkflowContextProperties.LAST_SCHEDULED_WORKFLOW.name(), workflow);
+    String lastScheduleWorkflow =
+        _record.getSimpleField(WorkflowContextProperties.LAST_SCHEDULED_WORKFLOW.name());
+    if (!workflow.equals(lastScheduleWorkflow)) {
+      _record.setSimpleField(WorkflowContextProperties.LAST_SCHEDULED_WORKFLOW.name(), workflow);
+      markWorkflowContextAsModified();
+    }
     // Record scheduled workflow into the history list as well
-    List<String> workflows = getScheduledWorkflows();
-    if (workflows == null) {
-      workflows = new ArrayList<String>();
-      _record.setListField(WorkflowContextProperties.SCHEDULED_WORKFLOWS.name(), workflows);
+    List<String> scheduledWorkflows = getScheduledWorkflows();
+    if (scheduledWorkflows == null) {
+      scheduledWorkflows = new ArrayList<String>();
+      _record.setListField(WorkflowContextProperties.SCHEDULED_WORKFLOWS.name(),
+          scheduledWorkflows);
+      markWorkflowContextAsModified();
+    }
+    if (!scheduledWorkflows.contains(workflow)) {
```

Comment:
Same here. Either you add the workflows in previous block. Or remove       markWorkflowContextAsModified() in previous one.

Reply:
Done.